src/transform/evalAssert.ml

```diff
@@ -0,0 +1,206 @@
+open Prelude
+open Cil
+open Formatcil
+module ES = SetDomain.Make(Exp.Exp)
+
+(** Instruments a program by inserting asserts either:
+    - After an assignment to a variable (unless trans.assert.full is activated) and
+    - At join points about all local variables
+
+                OR
+
+    - Only after pthread_mutex_lock (trans.assert.only-at-locks), about all locals and globals
+
+    Limitations without trans.assert.only-at locks:
+    - Currently only works for top-level variables (not inside an array, a struct, ...)
+    - Does not work for accesses through pointers
+    - At join points asserts all locals, but ideally should only assert ones that are
+        modified in one of the branches
+
+    Limitations in general:
+    - Removes comments, so if the original program had //UNKNOWN assertions, the annotation
+        will be removed and they will fail on the next iteration
+*)
+
+(* HACK to make the current fundec available to Apron when generating invariants, cannot use !MyCFG.current_node *)
+let currentFundec = ref dummyFunDec
+
+module EvalAssert = struct
+  (* should asserts of conjuncts be one-by-one instead of one big assert?  *)
+  let distinctAsserts = true
+
+  (* should asserts be surrounded by __VERIFIER_atomic_{begin,end}? *)
+  let surroundByAtomic = true
+
+  (* Cannot use Cilfacade.name_fundecs as assert() is external and has no fundec *)
+  let ass = ref (makeVarinfo true "assert" (TVoid []))
+  let atomicBegin = ref (makeVarinfo true "__VERIFIER_atomic_begin" (TVoid []))
+  let atomicEnd = ref (makeVarinfo true "__VERIFIER_atomic_end" (TVoid []))
```

Comment:
If these functions are already part of the program, we don't want a new `varinfo` for them, but instead set this to the `varinfo` we encountered while visiting `vglob`.

https://github.com/goblint/analyzer/blob/d1d175a89b889b3e4e7c440d9b2801a448de86e0/src/transform/evalAssert.ml#L66-L78

Reply:
I'm not sure what, if anything, the order of globals in a `Cil.file` guarantees, but it could be that you process some `GFun` _before_ you even see these declarations, so you end up using the created defaults anyway and mixing the two. Apparently we don't yet have a `Cilfacade.find_name_global`, but I guess we could.

Although, does it even matter which `varinfo` they use? This is just for printing at the end anyway, so couldn't you just always make them up and not even try to reuse?

Reply:
I went with always using fresh ones now